exporter/tanzuobservabilityexporter/config.go

```diff
@@ -42,32 +44,125 @@ type Config struct {
+// unchanged.
+func (c *Config) withoutMissingFields() (filledIn Config, err error) {
+	if c.Traces == nil && c.Metrics == nil {
+		return Config{}, errors.New("either the traces or metrics stanza must be present")
+	}
+	filledIn = *c
+	var tracesURL, metricsURL *url.URL
+	if filledIn.Metrics == nil {
+		tracesURL, err = parseEndpoint("traces", filledIn.Traces.Endpoint)
+		if err != nil {
+			return Config{}, err
+		}
+		metricsURL = &url.URL{}
+		*metricsURL = *tracesURL
+		metricsURL.Host = hostWithPort(tracesURL.Host, 2878)
+		filledIn.Metrics = &MetricsConfig{HTTPClientSettings: confighttp.HTTPClientSettings{Endpoint: metricsURL.String()}}
+	} else if filledIn.Traces == nil {
+		metricsURL, err = parseEndpoint("metrics", filledIn.Metrics.Endpoint)
+		if err != nil {
+			return Config{}, err
+		}
+		tracesURL = &url.URL{}
+		*tracesURL = *metricsURL
+		tracesURL.Host = hostWithPort(metricsURL.Host, 30001)
+		filledIn.Traces = &TracesConfig{HTTPClientSettings: confighttp.HTTPClientSettings{Endpoint: tracesURL.String()}}
+	} else {
+		metricsURL, err = parseEndpoint("metrics", filledIn.Metrics.Endpoint)
+		if err != nil {
+			return Config{}, err
+		}
+		tracesURL, err = parseEndpoint("traces", filledIn.Traces.Endpoint)
+		if err != nil {
+			return Config{}, err
+		}
+		if tracesURL.Hostname() != metricsURL.Hostname() {
+			return Config{}, errors.New("host for metrics and traces must be the same")
+		}
+	}
+	return
+}
+
+// sanitize changes this Config in place by filling in missing fields. If sanitize encounters an error,
+// it leaves this Config unchanged.
+func (c *Config) sanitize() error {
+	cfg, err := c.withoutMissingFields()
 	if err != nil {
 		return err
 	}
-	metricsURL, err := parseEndpoint("metrics", c.Metrics.Endpoint)
+	*c = cfg
+	return nil
+}
+
+// metricsHostNameAndPort returns the host name and port of the metrics endpoint.
+// metricsHostNameAndPort panics if it can't extract the host name and port.
+// Therefore, it should only be called after a successful call to sanitize(),
+// which guarantees a valid host name and port for the metrics endpoint.
+func (c *Config) metricsHostNameAndPort() (hostName string, port int) {
+	return hostNameAndPort(c.Metrics.Endpoint)
+}
+
+// tracesHostNameAndPort returns the host name and port of the traces endpoint.
+// metricsHostNameAndPort panics if it can't extract the host name and port.
+// Therefore, it should only be called after a successful call to sanitize(),
+// which guarantees a valid host name and port for the traces endpoint.
+func (c *Config) tracesHostNameAndPort() (hostName string, port int) {
+	return hostNameAndPort(c.Traces.Endpoint)
+}
+
+func hostNameAndPort(endpoint string) (hostName string, port int) {
+	u, err := url.Parse(endpoint)
+	if err != nil {
+		panic(err)
+	}
+	hostName = u.Hostname()
+	port, err = strconv.Atoi(u.Port())
 	if err != nil {
-		return err
+		panic(err)
```

Comment:
I am not confident this is good use of panic. Maybe someone more experienced with Go could confirm. I was anticipating returning the error instead

Reply:
I got rid of the panic and return the error instead.